### ISSUE TITLE
`multi_output_tree` training throughput collapses ~18x under multi-process parallelism

### ISSUE THREAD

**Issue by LL145** [dmlc/xgboost#12420]
## Summary

Since 3.4.0, training a `multi_strategy="multi_output_tree"` model with `nthread=1` from many concurrent single-process workers (the standard one-process-per-core hyperparameter-search pattern) collapses to a small constant aggregate throughput, regardless of how many worker processes run. On a 16C/32T machine, 3.3.0 sustains ~18 fits/s across 16 workers while 3.4.0 sustains ~1 fit/s at ~6% total CPU utilization. The machine sits mostly idle while every worker is blocked.

The regression is easiest to hit with shallow/strongly-regularized trees and a custom objective, which is a common configuration for noisy tabular data (finance etc.).

## Environment

- xgboost 3.4.0 (pip wheel) vs 3.3.0 (pip wheel) — same machine, same script
- Windows 11 Pro 10.0.26200
- AMD Ryzen 9 9950X (16 cores / 32 threads), 64 GB RAM
- Python 3.14.6 (MSC v.1944 64-bit), numpy 2.5.1, scipy 1.18.0

## Reproduction

Self-contained script, synthetic data. Each worker process trains with `nthread=1` on its own private copy of a small dataset, so N workers on an N-core machine should scale near-linearly.

```python
import os
import time
from concurrent.futures import ProcessPoolExecutor

WORKERS = int(os.environ.get("WORKERS", "16"))
FITS_PER_WORKER = 4
ROWS, FEATS, TARGETS, ROUNDS = 5000, 30, 6, 200

PARAMS = {
    "tree_method": "hist",
    "multi_strategy": "multi_output_tree",
    "grow_policy": "lossguide",
    "eta": 0.044,
    "alpha": 0.23,
    "lambda": 0.376,
    "gamma": 0.372,
    "min_child_weight": 2.07,
    "disable_default_eval_metric": 1,
    "nthread": 1,
}


def _make_data():
    import numpy as np
    rng = np.random.default_rng(0)
    X = rng.standard_normal((ROWS, FEATS))
    y = rng.standard_normal((ROWS, TARGETS)) * 0.01
    return X, y


def _custom_obj(pred, dtrain):
    # plain squared error written as a custom objective: the point is the
    # per-round python round-trip, not the loss itself
    import numpy as np
    y = dtrain.get_label().reshape(pred.shape)
    return pred - y, np.ones_like(pred)


def _fit_once(X, y):
    import xgboost
    dtrain = xgboost.DMatrix(X, label=y, nthread=1)
    xgboost.train(PARAMS, dtrain, num_boost_round=ROUNDS, obj=_custom_obj)


def _worker(_):
    X, y = _make_data()
    _fit_once(X, y)  # warm-up
    w0, c0 = time.perf_counter(), time.process_time()
    for _ in range(FITS_PER_WORKER):
        _fit_once(X, y)
    return time.perf_counter() - w0, time.process_time() - c0


if __name__ == "__main__":
    import xgboost

    X, y = _make_data()
    _fit_once(X, y)  # warm-up
    t0 = time.perf_counter()
    _fit_once(X, y)
    solo = time.perf_counter() - t0

    ncpu = os.cpu_count()
    with ProcessPoolExecutor(max_workers=WORKERS) as pool:
        w0 = time.perf_counter()
        results = list(pool.map(_worker, range(WORKERS)))
        wall = time.perf_counter() - w0

    total_cpu = sum(c for _, c in results)
    total_fits = WORKERS * FITS_PER_WORKER
    print(f"xgboost {xgboost.__version__}  ({ncpu} logical cpus, {WORKERS} workers)")
    print(f"  solo fit:              {solo:.2f}s")
    print(f"  parallel throughput:   {total_fits / wall:.2f} fits/s")
    print(f"  mean per-fit wall:     {sum(w for w, _ in results) / total_fits:.2f}s")
    print(f"  aggregate cpu use:     {total_cpu / (wall * ncpu) * 100:.0f}% of machine")
```

## Results

```
xgboost 3.3.0  (32 logical cpus, 16 workers)
  solo fit:              0.29s
  parallel throughput:   17.80 fits/s
  mean per-fit wall:     0.38s     <- barely degraded vs solo
  aggregate cpu use:     21% of machine

xgboost 3.4.0  (32 logical cpus, 16 workers)
  solo fit:              1.00s     <- 3.4x slower solo, too
  parallel throughput:   0.98 fits/s   <- 18x lower
  mean per-fit wall:     12.73s    <- 12.7x degradation under concurrency
  aggregate cpu use:     6% of machine
```

## Diagnostics gathered

- **Aggregate throughput is constant regardless of worker count** on 3.4.0: 8 workers → ~1.0 fits/s, 32 workers → ~0.95 fits/s (per-fit wall scales linearly with worker count: 6.2s at 8 workers, 25.5s at 32). It behaves like a system-wide serialized resource.
- **Context switches: ~560,000/s** system-wide during the 32-worker run (normal for this box is 10-50k), with kernel CPU time exceeding user CPU time in every worker (sampled 10s window: sum over 32 workers = 11s user + 18.2s kernel out of 320 possible CPU-seconds). The workers are asleep ~90% of the time; per-core load concentrates on a few cores that keep shuffling.
- **Thread churn**: a single fit-loop process with `nthread=1` holds a fluctuating ~82-113 threads on 3.4.0 (threads created/destroyed continuously during training). On 3.3.0 the same loop holds a stable 80. Passing `nthread=1` explicitly to the `DMatrix` constructor removes the churn and brings solo cpu/wall back to ~1.0, but does **not** fix the parallel collapse.
- **`OMP_NUM_THREADS=1` has no effect** on the collapse (still ~1 fit/s, ~7% CPU).
- A **built-in objective** (`reg:squarederror`, same `multi_output_tree` + hist + `nthread=1`, real dataset) also degrades under concurrency (26% CPU), though less than the custom-objective path (7-11% CPU). With default params on unregularized synthetic noise (deep trees) 3.4.0 actually outperforms 3.3.0 — the pathology needs shallow trees / low per-region work, where per-round overhead dominates.
- Both wheels link the same OpenMP runtime (`VCOMP140.DLL`); it is not a runtime swap. Memory is not a factor (dataset is ~1 MB/worker, 52 GB free).

## Suspicion

The 3.4.0 vector-leaf rework (batched split application / shared histogram building, e.g. #12365, #12397, #12387) appears to enter many more, much smaller OpenMP parallel regions per boosting round than 3.3.0, and at least some of those regions use the default team size instead of the booster's `nthread`, waking a full 32-thread team for microseconds of work. One process doing this is measurable (thread churn, cpu/wall 1.6 with nthread=1); N processes doing it concurrently collapse the scheduler (560k context switches/s) and the machine goes idle waiting on wakeups. The per-round custom-objective round-trip (`get_label`/prediction copy) seems to multiply the number of such regions.

Happy to run further diagnostics (ETW traces, specific builds) if useful.

**Comment by trivialfis** [comment]
I haven't been able to reproduce on Linux yet, this might be related to https://github.com/dmlc/xgboost/issues/11787 (the slow machine I mentioned there was Windows).

> The 3.4.0 vector-leaf rework (batched split application / shared histogram building

None of the PR you mentioned have anything to do with omp regions. #12397 is a Python demo test!

**Comment by trivialfis** [comment]
>     "min_child_weight": 2.07,

This is only supported on 3.4, which might also be a cause.

**Comment by LL145** [comment]
> > ```
> > "min_child_weight": 2.07,
> > ```
> 
> This is only supported on 3.4, which might also be a cause.



Thanks for taking a look !

I tested both hypotheses on the same Windows machine. Neither min_child_weight nor lossguide is the trigger — removing either changes nothing:


|variant                                            |3.3.0                               |3.4.0                               |
|---------------------------------------------------|------------------------------------|------------------------------------|
|as reported (`lossguide` + `min_child_weight=2.07`)|solo 0.30s, parallel **19.6 fits/s**|solo 1.00s, parallel **0.98 fits/s**|
|without `min_child_weight`                         |—                                   |solo 0.99s, **0.98 fits/s**         |
|`grow_policy="depthwise"` (with `min_child_weight`)|solo 0.29s, **20.0 fits/s**         |solo 0.99s, **0.99 fits/s**         |

**Comment by trivialfis** [comment]
I tested your script using a Windows laptop and couldn't reproduce the issue either:
```
xgboost 3.3.0  (16 logical cpus, 16 workers)
  solo fit:              0.66s
  parallel throughput:   4.54 fits/s
  mean per-fit wall:     2.08s
  aggregate cpu use:     55% of machine
```

```
xgboost 3.4.0  (16 logical cpus, 16 workers)
  solo fit:              0.40s
  parallel throughput:   7.24 fits/s
  mean per-fit wall:     0.98s
  aggregate cpu use:     42% of machine
```

The variance is quite high, but less than 30%, not many times over. I installed XGBoost through pypi:
```
Downloading xgboost-3.4.0-py3-none-win_amd64.whl (48.9 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.9/48.9 MB 30.8 MB/s  0:00:01
Installing collected packages: xgboost
Successfully installed xgboost-3.4.0
(xgboost_dev) C:\Users\jiamingy\ws\xgboost_dev\xgboost [master ≡]> cd ..\testing\
(xgboost_dev) C:\Users\jiamingy\ws\xgboost_dev\testing> python .\12420.py
```

**Comment by LL145** [comment]
@trivialfis 
Thank you for your reply!

Found the root cause 
3.4.0 is actually *faster*: `BoostOneIter` 0.32s vs 0.40s per 300 rounds.

It's the CUDA wheel. Every C API call goes through `XGBoostAPIGuard` → `cudaGetDevice()`. The 3.4.0 Windows wheel is built with CUDA 13, and on a machine with **no NVIDIA driver** (mine is AMD-only) its cudart seems to re-probe for the driver on every call instead of caching the failure — ~140µs of kernel time each time, and the probe path serializes across processes machine-wide.

```python
import time, numpy as np, xgboost
d = xgboost.DMatrix(np.zeros((10, 2)))
t0 = time.perf_counter()
for _ in range(2000):
    d.num_row()
print(f"{(time.perf_counter() - t0) / 2000 * 1e6:.1f} us/call")
```

0.23µs on 3.3.0, **139µs** on 3.4.0, 0.22µs on `xgboost-cpu` 3.4.0. A custom-objective round makes ~28 guarded API calls → ~3.9ms/round of probe tax, which matches the solo regression exactly (per-fit kernel time 0.03s → 0.77s, user time unchanged). And with the original repro script, `xgboost-cpu` 3.4.0 is the fastest of the three: solo 0.25s / 21.4 fits/s, vs 0.30s / 18.0 on 3.3.0 and 1.10s / 0.84 on the 3.4.0 CUDA wheel.

So that's why you couldn't reproduce it — your machines have NVIDIA drivers, so the probe succeeds and is cheap

**Comment by trivialfis** [comment]
Noted. Thank you for sharing. I need to find a way to disable the CUDA call. Closing since the original issue is resolved.
